### PR TITLE
Add Ledger hardware wallet support

### PR DESCRIPTION
**Pull request by TriPhan96**
Update Ledger device support
    
- Sign in with Ledger
- Sign transaction with Ledger: Transfer Tx, Modify Multisig Aggregate TX,
            Multi Sig. TX, Multisig TX, Provision Namespace TX,
            Mosaic Definition TX, Mosaic Supply Change
- Support delegated harvesting

### CODE REVIEW

File: build.js [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    app.on('ready', function () {
+      globalShortcut.register('CommandOrControl+R', function () {
+        // do nothing to forbidden default refresh
+      })
+    })
+  }
+  app.on('window-all-closed', function () {
+    app.quit()
+  })
+  // app.on('web-contents-created', (e, webContents) => {
```
Comment (rg911) [inline build.js]: Remove unused codes?

File: src/app/app.js [hunk 2]
```diff
@@ -122,11 +124,13 @@ window.$ = window.jQuery = $;
...
+require("babel-polyfill");
+require('babel-runtime/regenerator');
 
-// Apply nw.js config if using Chrome
-if (/Chrome/.test(navigator.userAgent) && /Google Inc/.test(navigator.vendor)) {
-    nwConfig();
-}
+// // Apply nw.js config if using Chrome
+// if (/Chrome/.test(navigator.userAgent) && /Google Inc/.test(navigator.vendor)) {
+//     nwConfig();
```
Comment (rg911) [inline src/app/app.js]: remove unused

File: src/app/app.js [hunk 3]
```diff
@@ -137,4 +141,4 @@ angular.module('app').run(appRun);
 
 angular.bootstrap(document, ['app'], {
     strictDi: true
-});
+});
```
Comment (rg911) [inline src/app/app.js]: lint issue, missing new line

File: src/app/modules/languages/en.js [hunk 4]
```diff
@@ -789,6 +791,23 @@ function EnglishProvider($translateProvider) {
         TREZOR_TEXT: 'The TREZOR hardware wallet allows you to secure your XEM, mosaics and multisig accounts',
         TREZOR_BUTTON: 'Login with TREZOR',
 
+        // NANO LEDGER RELATED
+        NANO_LEDGER_TITLE: 'NANO LEDGER',
+        NANO_LEDGER_TEXT: 'The NANO LEDGER hardware wallet allows you to secure your XEM, mosaics and multisig accounts',
+        NANO_LEDGER_BUTTON: 'Login with NANO LEDGER',
+        NANO_LEDGER_NO_DEVICE: 'Not found Ledger device',
```
Comment (rg911) [inline src/app/modules/languages/en.js]: Would probably say: `Ledger device not found`

File: src/app/modules/ledger/hw-app-nem.js [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            apdus.push(apdu);
+            offset += chunkSize;
+        }
+
+        let response = Buffer.alloc(0);
+        for (let apdu of apdus) {
+            response = await this.transport.send(apdu.cla, apdu.ins, apdu.p1, apdu.p2, apdu.data);
+        }
+        // if (response.toString() != CONTINUE_SENDING) {
+        //     console.log("handle ok")
```
Comment (rg911) [inline src/app/modules/ledger/hw-app-nem.js]: Remove comments

File: src/app/modules/ledger/ledger.controller.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        }
+    }
+
+    /**
+     * Login with LEDGER
+     */
+    async login() {
+        this.okPressed = true;
+        let checkVersion = await this._Ledger.getAppVersion();
+        if (checkVersion == 1) {
```
Comment (rg911) [inline src/app/modules/ledger/ledger.controller.js]: === 1

File: src/app/modules/ledger/ledger.service.js [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+            }))
+            .catch((err) => {
+                throw err;
+            });
+    }
+
+    bip44(network, index) {
+        // recognize networkId by bip32Path;
+        // "44'/43'/networkId'/walletIndex'/accountIndex'"
+        const networkId = network == -104 ? 152 : 104;
```
Comment (rg911) [inline src/app/modules/ledger/ledger.service.js]: Mijin (private) disabled?

File: src/app/modules/ledger/ledger.service.js [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+        }
+    }
+
+    async getRemoteAccount(hdKeypath) {
+        try {
+            const transport = await TransportNodeHid.open("");
+            const nemH = new NemH(transport);
+            try {
+                const result = await nemH.getRemoteAccount(hdKeypath)
+                transport.close();
```
Comment (rg911) [inline src/app/modules/ledger/ledger.service.js]: could probably put transport.close() into finally{}

File: src/app/modules/ledger/ledger.service.js [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+                        "hdKeypath": hdKeypath,
+                        "remoteAccount": result.remoteAccount
+                    }
+                );
+            } catch (err) {
+                transport.close();
+                throw err
+            }
+        } catch (err) {
+            if (err.statusCode != null) return Promise.reject(err.statusCode);
```
Comment (rg911) [inline src/app/modules/ledger/ledger.service.js]: A bit lint improvement: using `{`

File: src/app/modules/ledger/ledger.service.js [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+            else if (err.id != null) return Promise.reject(err.id);
```
Comment (rg911) [inline src/app/modules/ledger/ledger.service.js]: Could you elaborate on the logic here? Why reject err.id?

File: src/app/modules/ledger/ledger.service.js [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+                    });
+                }
+            });
+        });
+    }
+
+    serialize(transaction, account, symbolOptin) {
+        return new Promise(async (resolve, reject) => {
+            this.getAppVersion().then(async checkVersion => {
+                if (checkVersion == 1) {
```
Comment (rg911) [inline src/app/modules/ledger/ledger.service.js]: ===

File: src/app/modules/ledger/ledger.service.js [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+                    signature: sig.signature
+                }
+                return Promise.resolve(payload);
+            }
+            catch (err) {
+                transport.close();
+                throw err
+            }
+        } catch (err) {
+            if (err.statusCode != null) return Promise.resolve(err);
```
Comment (rg911) [inline src/app/modules/ledger/ledger.service.js]: not rejecting?

File: src/app/modules/ledger/ledger.service.js [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+                    alert("Please check your Ledger device!");
+                    this._$timeout(() => {
+                        this._Alert.ledgerFollowInstruction();
+                    });
+                    //Transaction with testnet and mainnet
+                    //Correct the signer
+                    transaction.signer = account.publicKey;
+
+                    //If it is a MosaicDefinition Creation Transaction, then correct the creator
+                    if (transaction.type == 0x4001) {
```
Comment (rg911) [inline src/app/modules/ledger/ledger.service.js:259]: Could you elaborate on why here to set the creator for mosaic?

Could could the serialization use `nem.model.transactions.prepare` in nem-sdk?

File: src/app/services/catapultOptin.service.js [hunk 14]
```diff
@@ -165,6 +168,50 @@ class CatapultOptin {
...
+                            return [JSON.stringify(serialized)];
+                        }
+                        return new Promise(resolve => setTimeout( () => signTransaction(i + 1).then((next) => {
+                            resolve([JSON.stringify(serialized)].concat(next));
+                        }).catch(err => resolve([null])), 500));
+                    }).catch(err => {
+                        let message;
+                        switch (err.statusCode) {
+                            case 27013:
+                                message = 'Signing Symbol Opt-in cancelled by user';
```
Comment (rg911) [inline src/app/services/catapultOptin.service.js]: missing translation?

File: src/app/services/catapultOptin.service.js [hunk 15]
```diff
@@ -165,6 +168,50 @@ class CatapultOptin {
...
+     */
+    _sendLedgerDTOs(common, dtos) {
+        return new Promise((resolve, reject) => {
+            Promise.all( dtos.map( dto => this.createTransactionFromDTO(dto, common))).then( transactions => {
+                const signTransaction = (i) => {
+                    return this._Ledger.serialize(transactions[i], this._Wallet.currentAccount, true).then( serialized => {
+                        if (transactions.length - 1 === i) {
+                            return [JSON.stringify(serialized)];
+                        }
+                        return new Promise(resolve => setTimeout( () => signTransaction(i + 1).then((next) => {
```
Comment (rg911) [inline src/app/services/catapultOptin.service.js]: could you elaborate on why set timer here?

File: src/app/services/voting.service.js [hunk 16]
```diff
@@ -124,7 +125,11 @@ class Voting {
 
         let account;
         if (common.isHW) {
-            account = new TrezorAccount(this._Wallet.currentAccount.address, this._Wallet.currentAccount.hdKeypath);
+            if (this._Wallet.algo == "trezor") {
+                account = new TrezorAccount(this._Wallet.currentAccount.address, this._Wallet.currentAccount.hdKeypath);
+            } else if (this._Wallet.algo == "ledger") {
+                account = {};
```
Comment (rg911) [inline src/app/services/voting.service.js]: not setting the current account for voting?

File: src/app/services/voting.service.js [hunk 17]
```diff
@@ -149,7 +154,15 @@ class Voting {
         const signTransaction = (i) => {
             let p;
             if (common.isHW) {
-                p = account.signTransaction(broadcastData.transactions[i]).first().toPromise();
+                if (this._Wallet.algo == "trezor") {
+                    p = account.signTransaction(broadcastData.transactions[i]).first().toPromise();
+                } else if (this._Wallet.algo == "ledger") {
+                    const transaction = broadcastData.transactions[i];
+                    transaction.signer = nem.PublicAccount.createWithPublicKey("462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce");
```
Comment (rg911) [inline src/app/services/voting.service.js:168]: Could you elaborate where is this publicKey comes from? `462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce`

File: src/app/services/voting.service.js [hunk 18]
```diff
@@ -161,7 +174,7 @@ class Voting {
                     return [signed].concat(next);
                 });
             }).catch(err => {
-                throw err;
+                throw undefined;
```
Comment (rg911) [inline src/app/services/voting.service.js]: why throw undefined?

File: src/app/services/voting.service.js [hunk 19]
```diff
@@ -318,14 +331,39 @@ class Voting {
...
         // sign
         let signedTransactionsPromise;
         if (common.isHW) {
-            signedTransactionsPromise = account.signSerialTransactions(votes).first().toPromise();
+            if (this._Wallet.algo == "trezor") {
+                signedTransactionsPromise = account.signSerialTransactions(votes).first().toPromise();
+            } else if (this._Wallet.algo == "ledger") {
+                const signTransaction = (i) => {
+                    const transaction = votes[i];
+                    transaction.signer = nem.PublicAccount.createWithPublicKey("462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce");
```
Comment (rg911) [inline src/app/services/voting.service.js:386]: same comments above

File: tests/specs/ledger.spec.js [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    let result = await LedgerService.serialize(transaction, account).catch(err => {
+      fail('Error while serializig transaction: ' + err);
+    });
+
+    // THEN
+    expect(LedgerService.signTransaction).toHaveBeenCalled();
+    expect(result).toEqual(expetedResult);
+    done();
+  });
+});
```
Comment (rg911) [inline tests/specs/ledger.spec.js]: Lint improvement, extra line

File: build.js (hunk 1 above)
Comment (TriPhan96) [reply build.js]: Removed

File: src/app/app.js (hunk 2 above)
Comment (TriPhan96) [reply src/app/app.js]: Removed

File: src/app/app.js (hunk 3 above)
Comment (TriPhan96) [reply src/app/app.js]: Added

File: src/app/modules/languages/en.js (hunk 4 above)
Comment (TriPhan96) [reply src/app/modules/languages/en.js]: Changed

File: src/app/modules/ledger/ledger.controller.js (hunk 6 above)
Comment (TriPhan96) [reply src/app/modules/ledger/ledger.controller.js]: Changed

File: src/app/modules/ledger/hw-app-nem.js (hunk 5 above)
Comment (TriPhan96) [reply src/app/modules/ledger/hw-app-nem.js]: Removed

File: src/app/modules/ledger/ledger.service.js (hunk 8 above)
Comment (TriPhan96) [reply src/app/modules/ledger/ledger.service.js]: Used finally{} in the code

File: src/app/modules/ledger/ledger.service.js (hunk 9 above)
Comment (TriPhan96) [reply src/app/modules/ledger/ledger.service.js]: Used {} to reformat the code

File: src/app/modules/ledger/ledger.service.js (hunk 11 above)
Comment (TriPhan96) [reply src/app/modules/ledger/ledger.service.js]: Changed to` ===`

File: src/app/modules/ledger/ledger.service.js (hunk 13 above)
Comment (TriPhan96) [reply src/app/modules/ledger/ledger.service.js:259]: Currently, the NEM-sdk hasn't support for hardware while preparing the transaction object yet when running the nem.model.transaction.prepare, for example when we create mosaic this function has no data of mosaic creator (ref [here](https://github.com/QuantumMechanics/NEM-sdk/blob/a9d93b1bced51c0c2c6838a01e233896da1dd847/src/model/transactions/mosaicDefinitionTransaction.js#L19-L21)) cause the wallet doesn't have private key of ledger device. So we have to correct the data of signer like [this](https://github.com/FDS-Ledger/NanoWallet/blob/d3734feceaa37fbc220317eb04d995fd29a62134/src/app/modules/ledger/ledger.service.js#L236-L242). Maybe some modification in NEM-sdk will fix it

File: src/app/services/voting.service.js (hunk 18 above)
Comment (TriPhan96) [reply src/app/services/voting.service.js]: Fixed, please have a look

File: src/app/modules/ledger/ledger.service.js (hunk 12 above)
Comment (TriPhan96) [reply src/app/modules/ledger/ledger.service.js]: Fixed, please have a look

File: src/app/modules/ledger/ledger.service.js (hunk 10 above)
Comment (TriPhan96) [reply src/app/modules/ledger/ledger.service.js]: Cause when not connected to Ledger device, rejected error object has only id field but not statusCode field

File: src/app/services/catapultOptin.service.js (hunk 15 above)
Comment (TriPhan96) [reply src/app/services/catapultOptin.service.js]: Removed timer

File: src/app/services/voting.service.js (hunk 16 above)
Comment (TriPhan96) [reply src/app/services/voting.service.js]: No need re-create the current for Ledger account to sign the transaction here

File: src/app/services/voting.service.js (hunk 17 above)
Comment (TriPhan96) [reply src/app/services/voting.service.js:168]: Currently, the NEM-sdk hasn't support for hardware while preparing the transaction object yet when running the nem.model.transaction.prepare, for example when we create mosaic this function has no data of mosaic creator (ref [here](https://github.com/QuantumMechanics/NEM-sdk/blob/a9d93b1bced51c0c2c6838a01e233896da1dd847/src/model/transactions/mosaicDefinitionTransaction.js#L19-L21)) cause the wallet doesn't have private key of ledger device. So we have to correct the data of signer like [this](https://github.com/FDS-Ledger/NanoWallet/blob/d3734feceaa37fbc220317eb04d995fd29a62134/src/app/modules/ledger/ledger.service.js#L236-L242)). Maybe some modification in NEM-sdk will fix it

File: src/app/services/voting.service.js (hunk 19 above)
Comment (TriPhan96) [reply src/app/services/voting.service.js:386]: Explained as above

File: tests/specs/ledger.spec.js (hunk 20 above)
Comment (TriPhan96) [reply tests/specs/ledger.spec.js]: Added

File: src/app/modules/languages/en.js (hunk 4 above)
Comment (TriPhan96) [reply src/app/modules/languages/en.js]: Changed

File: src/app/modules/ledger/ledger.service.js (hunk 7 above)
Comment (TriPhan96) [reply src/app/modules/ledger/ledger.service.js]: Change which will support for Mijin network

File: src/app/services/catapultOptin.service.js (hunk 14 above)
Comment (TriPhan96) [reply src/app/services/catapultOptin.service.js]: Changed, will support translation

**Review by rg911 — APPROVED** [review]

**Comment by QuantumMechanics** [comment]
Hi, 

I can't build the application to test and ensure all is fine without access to `catapult-optin-module` and `symbol-paper-wallets`.

Please grant me access to both, thank you.

File: src/app/services/voting.service.js (hunk 17 above)
Comment (QuantumMechanics) [reply src/app/services/voting.service.js:168]: Is it needed for Trezor though ? I don't get why it is different for it under Electron environment. `Trezor.serialize` does not correct the signer like `Ledger.serialize`.

Comment (TriPhan96) [reply src/app/services/voting.service.js:168]: 1. The voting NEM-service use [nem-voting](https://github.com/shierve/nem-voting/tree/master/src) to sign voting transactions. It uses a pop-up method to sign many transactions in 1 session which not work for Electron app (it keep the session open). Therefore we re-use the normal pop-up (sign only one transaction each time) to address this pop-up method for Electron build.
2. On Trezor, the signer is correct on [trezor-firmwalre](https://github.com/trezor/trezor-firmware/blob/7d8fbacba7c5945ca93068f6c44e0df134e6b657/crypto/nem.c#L367-L369). For Ledger use do that work on wallet before send it to Ledger devices.

Comment (QuantumMechanics) [reply src/app/services/voting.service.js:168]: Right got it, however the code in Trezor firmware correct the public key if `!signer` but `462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce` is used here, shouldn't it be empty then ?

**[EDIT]** I do see the same public key being used in [nem-trezor](https://github.com/shierve/nem-trezor/blob/master/src/trezor-account.ts#L31) but I can't determine where the correction happens with that wrong signer.

Comment (TriPhan96) [reply src/app/services/voting.service.js:168]: We just update signer for mosaic creation transaction (ref [here](https://github.com/FDS-Ledger/NanoWallet/blob/d3734feceaa37fbc220317eb04d995fd29a62134/src/app/modules/ledger/ledger.service.js#L236-L242)) so no signer correction for bot Trezor and Ledger wallets now. Do you mean that there's something wrong with voting signing and creating flow on hardware wallets?

Comment (QuantumMechanics) [reply src/app/services/voting.service.js:168]: > Do you mean that there's something wrong with voting signing and creating flow on hardware wallets?

Well it seems something is wrong with Trezor voting, a signer public key is set (`462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce`) but is not corrected maybe it is somewhere but I can't find it anywhere. The firmware correct the signer field if empty which is not the case here. I am completely lost between several files and libraries.

**Pull request merged**